rholang/src/test/scala/coop/rchain/rholang/ProcGen.scala

```diff
@@ -0,0 +1,247 @@
+        else new PNeg(new PGround(new GroundInt((-n).toString)))
+      })
+    lazy val groundBoolGen =
+      Gen.oneOf(new BoolFalse(), new BoolTrue()).map(b => new PGround(new GroundBool(b)))
+    lazy val groundStringGen =
+      Arbitrary.arbString.arbitrary.map(s => new PGround(new GroundString(withQuotes('"')(s))))
+    lazy val groundUriGen =
+      uriGen.map(s => new PGround(new GroundUri(s)))
+
+    Gen.oneOf(
+      groundIntGen,
+      groundBoolGen,
+      groundStringGen,
+      groundUriGen
+    )
+  }
+
+  private def nameQuoteGen(state: State): Gen[NameQuote] =
+    procGen(allProcs, state.decrementHeight).map(new NameQuote(_))
+  private def nameGen(state: State): Gen[Name] =
+    nameQuoteGen(state)
+
+  lazy val sendSingleGen: Gen[SendSingle]     = Gen.const(new SendSingle())
+  lazy val sendMultipleGen: Gen[SendMultiple] = Gen.const(new SendMultiple())
+  lazy val sendGen: Gen[Send] = Gen.oneOf(
+    sendSingleGen,
+    sendMultipleGen
+  )
+
+  private def psendGen(state: State): Gen[PSend] =
+    for {
+      name     <- nameGen(state.decrementHeight)
+      send     <- sendGen
+      listProc <- Gen.listOf(procGen(allProcs, state.decrementHeight))
+    } yield new PSend(name, send, seqToJavaCollection[ListProc, Proc](listProc))
+
+  lazy val pnilGen: Gen[PNil] = Gen.const(new PNil())
+
+  private def nameDeclSimplGen: Gen[NameDeclSimpl] = identifierGen.map(new NameDeclSimpl(_))
+
+  lazy val nameDeclUrnGen: Gen[NameDeclUrn] =
+    for {
+      name <- identifierGen
+      uri  <- uriGen
+    } yield new NameDeclUrn(name, uri)
+
+  lazy val nameDeclGen: Gen[NameDecl] = Gen.oneOf(nameDeclSimplGen, nameDeclUrnGen)
+
+  private def pnewGen(state: State): Gen[PNew] = {
+    val newState = state.decrementHeight
+
+    for {
+      seqNameDecl <- Gen.nonEmptyListOf(nameDeclSimplGen)
+      names       = extractNames(seqNameDecl)
+      proc        <- procGen(processContextProcs, newState.addNames(names))
+    } yield new PNew(seqToJavaCollection[ListNameDecl, NameDecl](seqNameDecl), proc)
+  }
+
+  lazy val bundleGen: Gen[Bundle] =
+    Gen.oneOf(new BundleWrite(), new BundleRead(), new BundleReadWrite(), new BundleEquiv())
+
+  private def procGen(procGens: Seq[State => Gen[Proc]], state: State): Gen[Proc] =
+    if (state.height > 0) {
+      oneOf(procGens.map(_.apply(state.decrementHeight)))
+    } else
+      pnilGen
+
+  private def allProcs: Seq[State => Gen[Proc]] =
+    Seq(pgroundGen, pparGen, psendGen, pnewGen, pvarGen)
+
+  private def processContextProcs: Seq[State => Gen[Proc]] =
+    Seq(pgroundGen, pparGen, psendGen, pnewGen)
+
+  def topLevelGen(height: Int): Gen[Proc] =
+    procGen(processContextProcs, State(height, Set.empty))
+
+  implicit def procShrinker: Shrink[Proc] = Shrink {
+    case p: PGround =>
+      (p.ground_ match {
+        case p: GroundInt =>
+          shrinkIntegral[Long]
+            .shrink(p.longliteral_.toLong)
+            .map(n => new GroundInt(n.toString))
```

Comment:
More of a curiosity, but:

If we had an instance of InvariantFunctor[Shrink] (which is really mechanic to write), we could do here:

```
val GroundIntShrink = 
  implicitly[Shrink[Long]].xmap(GroundInt(_.toString), _.longliteral.toLong)
GroundIntShrink.shrink(p)
```

https://leanpub.com/fpmortals/read#leanpub-auto-variance

In Cats lingo it'd be `Invariant` and `imap`. Funny how I mention this the second time within 2 hours in subsequently reviewed PRs :D https://github.com/rchain/rchain/pull/1700/files#r226205825

Reply:
sweet!